### PR TITLE
Use {:disable_nonlinear_arithmetic} to simplify standard library

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
- Use `{:disableNonlinearArithmetic}` on modules to simplify the definition and usage of the standard library
- Use `{:z3ArithmeticSolver 6}` on some modules to let their verification pass more reliably

### How has this been tested?
- This is a refactoring so no addition tests were needed

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

**Review by atomb — CHANGES_REQUESTED** [review]
I love this. The simplification is substantial. Before we merge it, though, we should figure out how to remove the `{:axiom}` annotations and make the other failing tests pass.

**Comment by keyboardDrummer** [comment]
> Before we merge it, though, we should figure out how to remove the `{:axiom}` annotations and make the other failing tests pass.

If the SMT solver's non linear arithmetic support is not doing the definitional proofs of the operators for us, then what else can we do but add definitional axioms for these operators?

**Comment by keyboardDrummer** [comment]
Comparing bruned Boogie files doesn't reveal any reason why the the NLA lemma's would suddenly fail.

Looking a the suddenly failing lemma `FundamentalDivMod`, and the pruned Boogie code that's sent to the solver.

Here it is with the change from this PR, causing the lemma not to verify:
[LemmaFundamentalDivModPrunedNoPass.txt](https://github.com/dafny-lang/dafny/files/13550306/LemmaFundamentalDivModPrunedNoPass.txt)

If I add `includes = ["**/*/ModInternalsNonlinear.dfy"]` to the `dfyconfig.toml`, I get an almost exactly the same pruned .bpl file, but verification will fail:
[LemmaFundamentalDivModPrunedPasses.txt](https://github.com/dafny-lang/dafny/files/13550307/LemmaFundamentalDivModPrunedPasses.txt)

The only difference is in the generated numbers.

**Comment by atomb** [comment]
This might be a case where it's worth exposing another use of the `:smt_option` attribute added to Boogie to support `:disable_nonlinear_arithmetic`. I've confirmed that all the definitions marked `{:axiom}` in your branch verify successfully with`smt.arith.solver=6`. It's something I hope most users wouldn't have to do themselves, but for the standard library it seems reasonable.

**Comment by keyboardDrummer** [comment]
> This might be a case where it's worth exposing another use of the `:smt_option` attribute added to Boogie to support `:disable_nonlinear_arithmetic`. I've confirmed that all the definitions marked `{:axiom}` in your branch verify successfully with`smt.arith.solver=6`. It's something I hope most users wouldn't have to do themselves, but for the standard library it seems reasonable.

Works, thanks

File: Source/DafnyCore/AST/Members/Method.cs [hunk 1]
```diff
@@ -46,7 +46,7 @@ public class Method : MemberDecl, TypeParameter.ParentType,
       yield return a;
     }
 
-    if (Body is null && HasPostcondition && EnclosingClass.EnclosingModuleDefinition.ModuleKind == ModuleKindEnum.Concrete && !HasExternAttribute) {
+    if (Body is null && HasPostcondition && EnclosingClass.EnclosingModuleDefinition.ModuleKind == ModuleKindEnum.Concrete && !HasExternAttribute && !HasAxiomAttribute) {
```
Comment (atomb) [inline Source/DafnyCore/AST/Members/Method.cs:49]: This change seems unrelated to the others. Was it intentionally included?

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/Arithmetic/Internal/DivInternalsNonlinear.dfy [hunk 2]
```diff
@@ -6,7 +6,14 @@
...
  *******************************************************************************/
 
-module DafnyStdLibs.Arithmetic.DivInternalsNonlinear {
+/**
+Several lemmas in this module are bodyless and marked as axiom.
+These lemmas could previously be proven with an empty body, but only if this file was verified by itself,
+not with the rest of the standard library.
+If we want to remove the {:axiom} annotation from these lemmas, 
+we will have to investigate why they only verify under specific conditions.   
+*/
```
Comment (atomb) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/Arithmetic/Internal/DivInternalsNonlinear.dfy:15]: This comment is obsolete now. You could comment on the arithmetic solver setting, though.

File: Source/DafnyStandardLibraries/src/DafnyStdLibs/Arithmetic/Internal/ModInternalsNonlinear.dfy [hunk 3]
```diff
@@ -6,7 +6,14 @@
...
  *******************************************************************************/
 
-module DafnyStdLibs.Arithmetic.ModInternalsNonlinear {
+/**
+Several lemmas in this module are bodyless and marked as axiom.
+These lemmas could previously be proven with an empty body, but only if this file was verified by itself,
+not with the rest of the standard library.
+If we want to remove the {:axiom} annotation from these lemmas, 
+we will have to investigate why they only verify under specific conditions.   
+*/
```
Comment (atomb) [inline Source/DafnyStandardLibraries/src/DafnyStdLibs/Arithmetic/Internal/ModInternalsNonlinear.dfy:15]: Another obsolete comment.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/verification/nonLinearArithmetic.dfy [hunk 4]
```diff
@@ -28,7 +28,7 @@ module Power0 {
 
 }
 
-module {:disable_nonlinear_arithmetic} Power1 {
+module {:disable-nonlinear-arithmetic} Power1 {
```
Comment (atomb) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/verification/nonLinearArithmetic.dfy:31]: I think this should be `disableNonlinearArithmetic`.

**Review by atomb — CHANGES_REQUESTED** [review]
Thanks for adding both new attributes! I noticed a few small inconsistencies and obsolete comments, but otherwise I think it looks good.

As to the final test failures, I've noticed that using `--boogie-filter "*LemmaModAuto*"` allows me to reproduce them locally (similarly for `DivAuto`). Both lemmas are kind of enormous. My inclination would be to take each of the conjuncts in the `DivAuto` and `ModAuto` predicates, prove each one in a completely separate lemma, and then call that collection of lemmas (and nothing else) from the top-level lemma.

File: Source/DafnyCore/AST/Members/Method.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyCore/AST/Members/Method.cs:49]: Yes, although it could be removed.

File: Source/IntegrationTests/TestFiles/LitTests/LitTest/auditor/TestAuditor.dfy-ietf.md.expect [hunk 5]
```diff
@@ -10,8 +10,6 @@
 
 * Declaration has explicit `{:axiom}` attribute. SHOULD provide a proof or test.
 
-* Ghost declaration has no body and has an ensures clause. MUST provide a body or add `{:axiom}`.
-
```
Comment (atomb) [inline Source/IntegrationTests/TestFiles/LitTests/LitTest/auditor/TestAuditor.dfy-ietf.md.expect:14]: This seems like something we want to continue reporting in the auditor.

Comment (keyboardDrummer) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/auditor/TestAuditor.dfy-ietf.md.expect:14]: But it has `{:axiom}` on it. The comment seemed incorrect.

Comment (atomb) [reply Source/IntegrationTests/TestFiles/LitTests/LitTest/auditor/TestAuditor.dfy-ietf.md.expect:14]: Ah! Got it. I didn't look at the source code it was reporting on. :)

**Review by atomb — DISMISSED** [review]

**Comment by atomb** [comment]
Looks like there are still some test failures.

**Review by robin-aws — APPROVED** [review]
Previously approved (but also wanted to say big thanks for this, it cleans up the source very nicely!)

**Comment by keyboardDrummer** [comment]
> Previously approved (but also wanted to say big thanks for this, it cleans up the source very nicely!)

Thanks!

Next up, getting rid of all the target specific .doo files
